hoodie-common/src/main/java/com/uber/hoodie/common/model/HoodieFileGroup.java

```diff
@@ -19,77 +19,66 @@
 package com.uber.hoodie.common.model;
 
 import com.uber.hoodie.common.table.HoodieTimeline;
-import com.uber.hoodie.common.table.timeline.HoodieInstant;
 import java.io.Serializable;
+import java.util.Collections;
 import java.util.Comparator;
 import java.util.List;
+import java.util.Map;
+import java.util.NavigableMap;
 import java.util.Optional;
 import java.util.TreeMap;
 import java.util.stream.Stream;
 
 /**
- * A set of data/base files + set of log files, that make up an unit for all operations
+ * A set of data/base files + set of log files with the same "file-id", that make up an unit for all operations.
+ *
+ * A file-group is identified by an unique file-Id within a partition. It has one or more file-slices.
+ *
+ * 3. Exactly one file-slice for a file-group provides the latest snapshot view (A view which is capable of providing
+ *    all the latest records in that file-group).
+ * 4. A file-group can have at-most one pending compaction.
+ * 5. The latest file-slice in a file-group will include all of
+ *    (a) the data-file corresponding to the last committed instant before a compaction request instant (if present)
+ *    (b) the set of log-files due to delta-instants after the instant (a) but before compaction request instant
+ *    (c) the set of log-files written after compaction request instant
  */
 public class HoodieFileGroup implements Serializable {
 
-  public static Comparator<String> getReverseCommitTimeComparator() {
-    return (o1, o2) -> {
-      // reverse the order
-      return o2.compareTo(o1);
-    };
-  }
-
   /**
    * Partition containing the file group.
    */
   private final String partitionPath;
-
   /**
    * uniquely identifies the file group
    */
   private final String id;
-
   /**
    * Slices of files in this group, sorted with greater commit first.
    */
-  private final TreeMap<String, FileSlice> fileSlices;
-
+  private final NavigableMap<String, FileSlice> fileSlices;
   /**
    * Timeline, based on which all getter work
    */
   private final HoodieTimeline timeline;
 
-  /**
-   * The last completed instant, that acts as a high watermark for all getters
-   */
-  private final Optional<HoodieInstant> lastInstant;
-
-  public HoodieFileGroup(String partitionPath, String id, HoodieTimeline timeline) {
+  public HoodieFileGroup(String partitionPath, String id, HoodieTimeline timeline,
```

Comment:
if we want to implement Builder pattern, lets do it the standard way by removing the public constructor and moving FileGroupBuilder inside FileGroup as an inner class ?